[stable/redis-ha] Upgrade chart to use stateful sets and pdbs

bzumhagen opened the pull request:
In order to fix issues brought up in https://github.com/kubernetes/charts/issues/3237.
It's a major version change because you cannot upgrade from deployment to stateful set without destroying the old deployment.

Stateful set fixes the issue of master selection thrashing, while adding the pds prevents cluster maintenance disruptions.

Also added redis-cli healthcheck to ensure redis service has at least started before moving on to creating more nodes.

Change-Id: Id6d296d8e49db925061e961a8e10988720dde3f2

*****************************************************************************************

Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the 
history. This will make it easier to identify new changes. The PR will be squashed 
anyways when it is merged. Thanks.

*****************************************************************************************

bzumhagen committed:
Upgrade chart to use stateful sets and pdbs

In order to fix issues brought up in https://github.com/kubernetes/charts/issues/3237.
It's a major version change because you cannot upgrade from deployment to stateful set without destroying the old deployment.

Stateful set fixes the issue of master selection thrashing, while adding the pds prevents cluster maintenance disruptions.

Change-Id: Id6d296d8e49db925061e961a8e10988720dde3f2

bzumhagen commented:
/assign @seanknox 
/assign @smileisak

unguiculus commented:
@bzumhagen There's a discussion in Slack about current problems with the chart. Would you mind joining it?

https://kubernetes.slack.com/archives/C6E3XH1ED/p1520362521000241?thread_ts=1520362521.000241&cid=C6E3XH1ED

unguiculus commented:
/assign

bzumhagen commented:
@unguiculus I don't seem to have a kubernetes slack account. How can I get an invite?

j-kaplan commented:
@bzumhagen you can sign up at http://slack.k8s.io/

bzumhagen committed:
Change rbac to match best practices

Change-Id: Ie1729820bccd1b0e8064358ffd6f81b2ea388465

unguiculus commented on stable/redis-ha/templates/redis-role.yaml:
```diff
@@ -1,20 +1,20 @@
-{{- if .Values.rbac.create -}}
```
This was correct.

bzumhagen replied:
Done

scottrigby replied:
@bzumhagen each line in this file is prepended with a `-`. Assuming this was accidental?

unguiculus commented on stable/redis-ha/templates/redis-rolebinding.yaml:
```diff
@@ -1,4 +1,4 @@
-{{- if .Values.rbac.create -}}
```
This was correct

bzumhagen replied:
Done

unguiculus commented on stable/redis-ha/templates/redis-sentinel-deployment.yaml:
```diff
@@ -11,7 +11,9 @@ spec:
         name: {{ template "fullname" . }}-sentinel
 {{ include "labels.standard" . | indent 8 }}
     spec:
+      {{ if .Values.serviceAccount.create }}
```
No need to warp this into an `if` block

bzumhagen replied:
Done

unguiculus commented on stable/redis-ha/templates/redis-server-statefulset.yaml:
```diff
@@ -19,7 +34,9 @@ spec:
         redis-node: "true"
 {{ include "labels.standard" . | indent 8 }}
     spec:
+      {{ if .Values.serviceAccount.create }}
```
No need to warp this into an `if` block

bzumhagen replied:
Done

unguiculus commented on stable/redis-ha/values.yaml:
```diff
@@ -31,13 +31,12 @@ servers:
   serviceType: ClusterIP  # [ClusterIP|LoadBalancer]
   annotations: {}
 
-rbac:
```
Leave this in

bzumhagen replied:
Done

unguiculus reviewed (changes requested):
I did a quick review of the RBAC changes. Please have a look at https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/rbac.md

bzumhagen committed:
Fix rbac values

bzumhagen commented:
Made requested changes

listingmirror commented:
Not sure this is the right place for this, but a HA chart with no node affinity seems bad. I tried the current redis-ha chart, and my first deploy put all 3 pods and 3 sentinels on the same node. *sigh*

j-kaplan commented:
@unguiculus What is the status of this? Looks like @bzumhagen has made the requested changes. Can we get another review? Thanks!

timstoop commented:
@bzumhagen Could you merge from master please? (not rebase)

MichaelSp commented:
@bzumhagen would you please consider this PR https://github.com/bzumhagen/docker-images/pull/5 and add an option to enable AUTH.
https://github.com/kubernetes/charts/blob/master/stable/redis/templates/secrets.yaml

unguiculus commented:
Please merge in the latest master and I'll be happy to review.

thuandt commented:
@bzumhagen I think we should use persistent volume for statefulset using `volumeClaimTemplates`

```yaml
{{- if .Values.persistentVolume.enabled }}
  volumeClaimTemplates:
    - metadata:
        name: data
        annotations:
        {{- range $key, $value := .Values.persistentVolume.annotations }}
          {{ $key }}: {{ $value }}
        {{- end }}
      spec:
        accessModes:
        {{- range .Values.persistentVolume.accessModes }}
          - {{ . | quote }}
        {{- end }}
        resources:
          requests:
            storage: {{ .Values.persistentVolume.size | quote }}
      {{- if .Values.persistentVolume.storageClass }}
      {{- if (eq "-" .Values.persistentVolume.storageClass) }}
        storageClassName: ""
      {{- else }}
        storageClassName: "{{ .Values.persistentVolume.storageClass }}"
      {{- end }}
      {{- end }}
{{- else }}
      volumes:
        - name: data
          emptyDir: {}
{{- end }}

```

with values

```yaml
persistentVolume:
  enabled: false
  # storageClass: "-"
  accessModes:
    - ReadWriteOnce
  size: 8Gi
  annotations: {}

```

@bzumhagen thanks for your work in this chart, it works really well in my cluster

scottrigby commented on stable/redis-ha/templates/redis-rolebinding.yaml:
```diff
@@ -15,4 +15,4 @@ roleRef:
 - kind: ServiceAccount
   name: {{ template "redis-ha.serviceAccountName" . }}
-{{- end -}}
+{{- end -}}
```
Newline at EOF should not be removed

scottrigby commented on stable/redis-ha/templates/redis-sentinel-pdb.yaml:
```diff
@@ -0,0 +1,11 @@
+    matchLabels:
+      app: "redis-ha"
+      redis-role: "sentinel"
+      release: "{{ .Release.Name }}"
```
Needs newline at EOF

scottrigby commented on stable/redis-ha/templates/redis-server-pdb.yaml:
```diff
@@ -0,0 +1,11 @@
+    matchLabels:
+      app: "redis-ha"
+      redis-node: "true"
+      release: "{{ .Release.Name }}"
```
Needs newline at EOF

scottrigby commented on stable/redis-ha/templates/redis-server-statefulset.yaml:
```diff
@@ -31,18 +48,24 @@ spec:
       volumes:
-      - name: data
-
+      - name: data
```
Needs newline at EOF

bzumhagen replied:
done

scottrigby commented on stable/redis-ha/Chart.yaml:
```diff
@@ -5,17 +5,19 @@ keywords:
+- email: ianmaddox@google.com
+  name: ianmaddox
+- email: bzumhagen@gmail.com
+  name: Ben Zumhagen
```
@bzumhagen Please also add an [OWNERS](https://github.com/kubernetes/charts#owning-and-maintaining-a-chart) file while you're at it

bzumhagen replied:
Done

scottrigby commented on stable/redis-ha/README.md:
```diff
@@ -58,28 +58,28 @@ $ helm install \
+| `tolerations`                    | Toleration labels for pod assignment                  | []                                                        |
+| `servers.serviceType`            | Set to "LoadBalancer" to enable access from the VPC   | ClusterIP                                                 |
+| `servers.annotations`            | See Appliance mode                                    | ``                                                        |
+| `serviceAccount.create`                    | If true, create & use RBAC resources                  | `false`                                                   |
```
If the goal is to keep the table aligned, `serviceAccount.create` is out of line.

scottrigby commented on stable/redis-ha/README.md:
```diff
@@ -58,28 +58,28 @@ $ helm install \
+| Parameter                        | Description                                           | Default                                                   |
+| -------------------------------- | ----------------------------------------------------- | --------------------------------------------------------- |
+| `redis_image`                    | Redis image                                           | `quay.io/smile/redis:4.0.6r2`                             |
+| `resources.server`               | CPU/Memory for redis nodes resource requests/limits  | Memory: `200Mi`, CPU: `100m`                              |
```
If the goal is to keep the table aligned, `resources.server` is out of line.

scottrigby commented on stable/redis-ha/README.md:
```diff
@@ -58,28 +58,28 @@ $ helm install \
+| `servers.serviceType`            | Set to "LoadBalancer" to enable access from the VPC   | ClusterIP                                                 |
+| `servers.annotations`            | See Appliance mode                                    | ``                                                        |
+| `serviceAccount.create`                    | If true, create & use RBAC resources                  | `false`                                                   |
+| `max_memory`                     | Sets the maximum memory redis can allocate            | "" This defaults to the maximum memory allowed               |
```
- `max_memory` is out of line with the table
- Also, values keys should be lower camel case (`maxMemory`). See [Chart Best Practices > Naming Conventions](https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/values.md#naming-conventions).
    - I see this chart already had a mix of values key case patterns (for example, there is both `redis_image` and `nodeSelector`). You could fix that in this PR as well (change `redis_image` to `redisImage` or just `image`).

bzumhagen committed:
Add additional changes to further stabilze chart

1. Sentinels are now using stateful set
2. Updating to latest redis image
3. Added refresh sentinels script to image to allow for better recovery
4. Tweaked pdb settings
5. Optimized labels

Change-Id: Id9acbfe1a65472cfccafeb08c676f6bbbe300f59

bzumhagen commented:
Hey guys, sorry for the delay. We discovered some issues with the chart and were trying to iron them out in our cluster.

I've rebased and added our latest findings, so hopefully this means a very stable chart which can survive cluster maintenance.

timstoop commented on stable/redis-ha/templates/redis-sentinel-pdb.yaml:
```diff
@@ -0,0 +1,10 @@
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: {{ template "redis-ha.fullname" . }}-sentinel-pdb
```
Why not add the default labels here as well?

bzumhagen replied:
Good catch

timstoop commented on stable/redis-ha/templates/redis-server-pdb.yaml:
```diff
@@ -0,0 +1,10 @@
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: {{ template "redis-ha.fullname" . }}-server-pdb
```
Why not add the default labels here as well?

bzumhagen replied:
Will do

timstoop commented on stable/redis-ha/templates/redis-server-statefulset.yaml:
```diff
@@ -1,5 +1,20 @@
+metadata:
+  name: {{ template "redis-ha.fullname" . }}-server-svc
+  labels:
+    chart: {{ template "chartref" . }}
```
Why not use the "labels.standard" here as well? As you do that for the Sentinel server, I'd expect it here as well.

bzumhagen replied:
Done

timstoop commented on stable/redis-ha/README.md:
```diff
@@ -58,28 +58,31 @@ $ helm install \
+| `servers.serviceType`            | Set to "LoadBalancer" to enable access from the VPC   | ClusterIP                                                 |
+| `servers.annotations`            | See Appliance mode                                    | ``                                                        |
+| `serviceAccount.create`                    | If true, create & use RBAC resources        | `false`                                                   |
+| `redis.max_memory`               | Sets the maximum memory redis can allocate            | "" (This defaults to the maximum memory allowed)          |
```
I tend to agree with an earlier comment from @scottrigby, please change the new keys to lower camel case (so it would be `redis.maxMemory`) as described in the Best Practises: https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/values.md#naming-conventions

bzumhagen replied:
sounds good

timstoop reviewed (changes requested):
Looking good! Just a few minor questions and change requests.

timstoop commented on stable/redis-ha/templates/redis-server-statefulset.yaml:
```diff
@@ -8,17 +23,20 @@ metadata:
     spec:
-      serviceAccountName: {{ template "redis-ha.serviceAccountName" . }}
+      {{ if .Values.rbac.create }}
+      serviceAccountName: {{ template "redis-ha.fullname" . }}-svcacct
```
Does this work for you with the values as given in values.yaml? As when I try to deploy this, the ServiceAccount that's created gets the name `idolized-lemur-redis-ha`, without the `-svcacct` added. When I look at the definition in _helpers.tpl, it seems that the template "redis-ha.serviceAccountName" does not have this `svcacct` part.

bzumhagen replied:
You are right. I think this was a rebase mistake

ghost commented:
Great effort on making the chart really usable.
One thing I noticed after deploying the latest version is from the sentinel pod logs:

`redis-ha2-redis-ha-sentinel-1 sentinel sh: refresh-sentinel.sh: not found
`
It keeps on repeating from all pods.

bzumhagen commented:
@tomhau01 I need to make some tweaks as per @timstoop's feedback. Looks like I messed some stuff up when doing my rebase :)

thuandt commented:
@bzumhagen in your redis docker image Dockerfile, you forgot add `refresh-sentinel.sh` to `/usr/local/bin`

bzumhagen committed:
Merge branch 'master' into master

bzumhagen committed:
Updating docker image

Latest docker image includes a fix for missing script

Change-Id: I71099850044a989e1782daf95380f5addcb309e9

bzumhagen commented:
I think i've addressed the outstanding comments. Let's give this another go.

unguiculus commented on stable/redis-ha/Chart.yaml:
```diff
@@ -5,19 +5,20 @@ keywords:
+- email: ianmaddox@google.com
+  name: ianmaddox
+- email: bzumhagen@gmail.com
+  name: Ben Zumhagen
```
Use your Github username.

unguiculus commented on stable/redis-ha/templates/redis-sentinel-pdb.yaml:
```diff
@@ -0,0 +1,10 @@
+  selector:
+    matchLabels:
+      redis-role: "sentinel"
+{{ include "labels.standard" . | indent 6 }}
```
Selectors should not have `chart` and `heritage` labels.

unguiculus commented on stable/redis-ha/templates/redis-sentinel-statefulset.yaml:
```diff
@@ -0,0 +1,66 @@
+      targetPort: 26379
+  selector:
+    redis-role: "sentinel"
+{{ include "labels.standard" . | indent 4 }}
```
Selectors should not have `chart` and `heritage` labels.

unguiculus commented on stable/redis-ha/templates/redis-sentinel-statefulset.yaml:
```diff
@@ -0,0 +1,66 @@
+  selector:
+    redis-role: "sentinel"
+{{ include "labels.standard" . | indent 4 }}
+---
```
Create separate files per resource.

unguiculus commented on stable/redis-ha/templates/redis-server-pdb.yaml:
```diff
@@ -0,0 +1,12 @@
+  selector:
+    matchLabels:
+      redis-node: "true"
+{{ include "labels.standard" . | indent 6 }}
```
Selectors should not have `chart` and `heritage` labels.

unguiculus commented on stable/redis-ha/templates/redis-server-statefulset.yaml:
```diff
@@ -1,5 +1,21 @@
+{{ include "labels.standard" . | indent 4 }}
+  type: "{{ .Values.servers.serviceType }}"
+  clusterIP: None
+---
```
Create separate files per resource

unguiculus commented on stable/redis-ha/templates/redis-master-service.yaml:
```diff
@@ -12,8 +13,7 @@ spec:
-    release: "{{ .Release.Name }}"
     redis-node: "true"
     redis-role: "master"
+{{ include "labels.standard" . | indent 4 }}
```
Selectors should not have `chart` and `heritage` labels.

unguiculus commented on stable/redis-ha/templates/redis-sentinel-statefulset.yaml:
```diff
@@ -0,0 +1,66 @@
+{{ include "labels.standard" . | indent 4 }}
+  annotations:
+{{ toYaml .Values.servers.annotations | indent 4 }}
+spec:
```
Governing services for statefulsets should be headless.

unguiculus commented on stable/redis-ha/templates/redis-server-statefulset.yaml:
```diff
@@ -1,5 +1,21 @@
+  selector:
+    redis-node: "true"
+{{ include "labels.standard" . | indent 4 }}
+  type: "{{ .Values.servers.serviceType }}"
```
Governing services for statefulsets should be headless, thus type should be hard-coded to `ClusterIP`.
